### PR TITLE
feat: get_certificates and get_csrs

### PR DESCRIPTION
**Pull request by Kayra1**
# Description

Added 2 new functions:
`get_assigned_certificates` gets all of the certificates that are available in the remote application databag.
`get_certificate_signing_requests` gets all of the generated certificate signing requests. It also provides the option to filter the csrs by wether they have their certificate available or not. 

## Checklist

- [x] My code follows the [style guidelines](/CONTRIBUTING.md) of this project.
- [x] I have performed a self-review of my own code.
- [ ] I have made corresponding changes to the documentation.
- [x] I have added tests that validate the behaviour of the software.
- [x] I validated that new and existing unit tests pass locally with my changes.
- [x] Any dependent changes have been merged and published in downstream modules.
- [x] I have bumped the version of any required library.

### CODE REVIEW

**Comment by Kayra1** [comment]
closes #88 and #89

File: lib/charms/tls_certificates_interface/v2/tls_certificates.py
Comment (saltiyazan) [inline lib/charms/tls_certificates_interface/v2/tls_certificates.py:1686]: There is a function called `get_requirer_csrs`, having both with these names make it a bit confusing.
I'm not saying we should only keep one, but we might want to rethink the names.

Comment (Kayra1) [reply lib/charms/tls_certificates_interface/v2/tls_certificates.py:1686]: That is in the Provider. This function is in the Requirer. I could change the name of the one in the requirer to be identical?

File: lib/charms/tls_certificates_interface/v2/tls_certificates.py
Comment (saltiyazan) [inline lib/charms/tls_certificates_interface/v2/tls_certificates.py:1688]: I'd provide that option too.

File: tests/unit/charms/tls_certificates_interface/v2/test_tls_certificates_v2_requires.py
Comment (saltiyazan) [inline tests/unit/charms/tls_certificates_interface/v2/test_tls_certificates_v2_requires.py:1247]: I think we should have separate test cases for different scenarios.

File: tests/unit/charms/tls_certificates_interface/v2/test_tls_certificates_v2_requires.py
Comment (saltiyazan) [inline tests/unit/charms/tls_certificates_interface/v2/test_tls_certificates_v2_requires.py:1149]: Should we have a case that asserts that only the certificates of this unit are returned in case we have more than one unit?
And maybe a case when an empty list is returned when there are no csrs/certficates?

File: lib/charms/tls_certificates_interface/v2/tls_certificates.py
Comment (Kayra1) [reply lib/charms/tls_certificates_interface/v2/tls_certificates.py:1688]: I've been investigating how to do this today, and I don't think a unit can read the databag of another unit of the same application unless it's a peer relation, making this impossible. Can you verify?

Comment (saltiyazan) [reply lib/charms/tls_certificates_interface/v2/tls_certificates.py]: Units can read each other's databags, I just don't know how to iterate over the units of an application, I have been reading a bit too but I didn't find a straightforward way. One could use the unit names like [here](_requirer_csrs), but I don't think that's what we are looking for here.

**Review by saltiyazan — COMMENTED** [review]
Looks good, but I'm interested what the others think regarding having one or three functions for `get_certificate_signing_requests`, I requested @gruyaume's review to get a second opinion on that.

File: lib/charms/tls_certificates_interface/v2/tls_certificates.py
Comment (gruyaume) [inline lib/charms/tls_certificates_interface/v2/tls_certificates.py:1024]: This comment can be next to the `csr` args a couple of lines lower

File: lib/charms/tls_certificates_interface/v2/tls_certificates.py
Comment (gruyaume) [reply lib/charms/tls_certificates_interface/v2/tls_certificates.py:1687]: If we have independent usecases for those 3 circumstances, I agree that 3 different methods makes sense. It's generally preferable to have smaller functions that have 1 purpose instead of multiple.

Let's only create the methods we know we need though. For example if we know we need 2 out of those 3, let's not implement the third one "just in case".

File: lib/charms/tls_certificates_interface/v2/tls_certificates.py
Comment (Kayra1) [reply lib/charms/tls_certificates_interface/v2/tls_certificates.py]: I've talked to the Juju team, and they told me that the preferred way is to look at the units of the peer relation. I will try to see if I can use that.

File: lib/charms/tls_certificates_interface/v2/tls_certificates.py
Comment (Kayra1) [reply lib/charms/tls_certificates_interface/v2/tls_certificates.py]: Done

File: lib/charms/tls_certificates_interface/v2/tls_certificates.py
Comment (gruyaume) [reply lib/charms/tls_certificates_interface/v2/tls_certificates.py:1685]: I don't have a very strong opinion on this, I'll let you make the call @Kayra1 as you're right in it at the moment.

File: lib/charms/tls_certificates_interface/v2/tls_certificates.py
Comment (gruyaume) [reply lib/charms/tls_certificates_interface/v2/tls_certificates.py]: Agreed, if we don't need it, let's not write code for it just in case.

File: lib/charms/tls_certificates_interface/v2/tls_certificates.py
Comment (Kayra1) [reply lib/charms/tls_certificates_interface/v2/tls_certificates.py:1685]: This is fine

File: lib/charms/tls_certificates_interface/v2/tls_certificates.py
Comment (saltiyazan) [inline lib/charms/tls_certificates_interface/v2/tls_certificates.py:1685]: Could you please remove these comments before merging?
I think we have a conclusion on all, right?

Comment (Kayra1) [reply lib/charms/tls_certificates_interface/v2/tls_certificates.py]: I was just working on it. Removed.

**Comment by Kayra1** [comment]
I also added a new feature called `get_expiring_certificates` and two tests along with it, there is a need for it right now. Please review that also.

File: lib/charms/tls_certificates_interface/v2/tls_certificates.py
Comment (saltiyazan) [inline lib/charms/tls_certificates_interface/v2/tls_certificates.py:1717]: I think if a user misuses this function and sets both to True, they will get an empty list.
I don't know if we want to consider that case or not :)

File: tests/unit/charms/tls_certificates_interface/v2/test_tls_certificates_v2_requires.py
Comment (saltiyazan) [inline tests/unit/charms/tls_certificates_interface/v2/test_tls_certificates_v2_requires.py:1240]: Is this needed here?

File: tests/unit/charms/tls_certificates_interface/v2/test_tls_certificates_v2_requires.py
Comment (saltiyazan) [inline tests/unit/charms/tls_certificates_interface/v2/test_tls_certificates_v2_requires.py:1265]: Is this needed?

File: tests/unit/charms/tls_certificates_interface/v2/test_tls_certificates_v2_requires.py
Comment (saltiyazan) [reply tests/unit/charms/tls_certificates_interface/v2/test_tls_certificates_v2_requires.py:1240]: It is, never mind

File: lib/charms/tls_certificates_interface/v2/tls_certificates.py
Comment (saltiyazan) [reply lib/charms/tls_certificates_interface/v2/tls_certificates.py:1717]: I see you have added a test case for this.
I think that could be enough.

**Review by saltiyazan — APPROVED** [review]

File: lib/charms/tls_certificates_interface/v2/tls_certificates.py
Comment (Kayra1) [reply lib/charms/tls_certificates_interface/v2/tls_certificates.py:1717]: There is a test specifically verifying that this happens :D. This makes sense in both an English way, Python way and a logical way, so I think it's ok.

**Pull request merged**
